prevent empty datafiles from being made + celery monthly task added to schedule

gordonmslai opened the pull request.

gordonmslai commented:
ValidationError appropriate?

aschn commented on watttime_grid_api/apps/datafile/tasks.py:
```diff
@@ -27,3 +29,13 @@ def csv_create(ba, market, year, month):
+            try:
+                csv_create(ba, market_abbrev, year, month)
+            except ValidationError:
+                print "No data for " + ba.abbrev + "_" + market_abbrev + "_" + str(year) + "_" + str(month) + ". Did not create DataFile/csv"
```
It ends up working better to use `logger.info` instead of `print` for messages like this in production

aschn commented on watttime_grid_api/apps/datafile/models.py:
```diff
@@ -34,6 +36,14 @@ class DataFile(models.Model):
     data_file = models.FileField()
 
+    def save(self, *args, **kwargs):
+        try:
```
Can we move this check to `generate_csv` and/or `save_csv` instead of `save`? I don't think `save` actually creates the csv file, so it could be reasonable to create a DataFile object then associate it with DataPoints after, then make sure to check when we actually create the csv.

gordonmslai replied:
Yep we can definitely do that! I thought we wanted to prevent a DataFile object from being created period (if it would be empty), and this does just that -- raises an error before a DataFile gets written into the database so that there's no possibility of even trying to create an empty csv. Let me know!

aschn replied:
It makes more sense to me to allow the row in the database to be created
but not to allow its associated file to be saved, but in practice it will
work pretty much the same either way so it's not a big deal. To clarify,
maybe make tests for both scenarios separately so we can be super explicit?

On Fri, Jul 31, 2015 at 11:14 AM, gordonmslai <notifications@github.com>
wrote:

> In watttime_grid_api/apps/datafile/models.py
> <https://github.com/WattTime/watttime-grid-api/pull/40#discussion_r36000092>
> :
>
> > @@ -34,6 +36,14 @@ class DataFile(models.Model):
> >      # the datafile
> >      data_file = models.FileField()
> >
> > +    def save(self, *args, **kwargs):
> > +        try:
>
> Yep we can definitely do that! I thought we wanted to prevent a DataFile
> object from being created period (if it would be empty), and this does just
> that -- raises an error before a DataFile gets written into the database so
> that there's no possibility of even trying to create an empty csv. Let me
> know!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WattTime/watttime-grid-api/pull/40/files#r36000092>.
>

gordonmslai replied:
Ok then I'll just move the raise Error to save_csv and take out what I have for save!

aschn commented on watttime_grid_api/apps/etl/schedule.py:
```diff
@@ -78,6 +78,12 @@
+    'create-last-month-csv': {
+        'task': 'apps.datafile.tasks.csv_create_all',
+        'schedule': crontab(0, 0, day_of_month='1'),
+        'args': [datetime.now().year, (datetime.now() - timedelta(days=1)).month],
```
A couple comments:
1) This raises an error for me because celery doesn't let you pass non-JSON-serializable objects as args to tasks. It also assumes that the code to generate the args is executed every time, which seems totally reasonable but I don't think is actually how celery works. What I usually do instead is something like:
```
def task(current_time=None):
    if not current_time:
        current_time = now()
```
2) When choosing the schedule to run the task, I'm not sure if I mentioned that the schedule uses UTC. So make sure to choose a time that works for local timezones... Maybe safer to make it on the 2nd day of the month?

gordonmslai replied:
Ok sure so I'll change the scheduled day of month and incorporate the datetime.now into the task  itself! I was wondering about that too -- wasn't able to set up celery on my computer to test it.

aschn replied:
Hmm yeah celery is tricky enough to set up on Mac, I bet it's a nightmare
on Windows.

On Fri, Jul 31, 2015 at 11:17 AM, gordonmslai <notifications@github.com>
wrote:

> In watttime_grid_api/apps/etl/schedule.py
> <https://github.com/WattTime/watttime-grid-api/pull/40#discussion_r36000469>
> :
>
> > @@ -78,6 +78,12 @@
> >
> >  # other tasks
> >  schedule.update({
> > +    # csv create for previous month
> > +    'create-last-month-csv': {
> > +        'task': 'apps.datafile.tasks.csv_create_all',
> > +        'schedule': crontab(0, 0, day_of_month='1'),
> > +        'args': [datetime.now().year, (datetime.now() - timedelta(days=1)).month],
>
> Ok sure so I'll change the scheduled day of month and incorporate the
> datetime.now into the task itself! I was wondering about that too -- wasn't
> able to set up celery on my computer to test it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/WattTime/watttime-grid-api/pull/40/files#r36000469>.
>

aschn commented:
Good stuff! I just added a few comments for things I should have been more clear about to begin with :)

gordonmslai commented:
Looks like im going to have to fix the downloads tests since the page should now prevent those empty datafiles from being shown. Any clue about the moer test failure in schedule?

gordonmslai commented:
turns out the tests were fine, except that because its the 31st, the Setup made datafiles in different months so it messed with the assertEquals. Changed "now" to the 15th. Also added a tearDown to delete all datafiles from the test storage. Forgot about that and a lot of files were accumulating.

aschn commented:
I pushed a fix to the schedule test, hope that works for you. Let's make sure to review the "files changed" view before approving this PR because there's been a lot of merging!

gordonmslai commented:
Were you able to take a look at this Anna?

aschn commented on watttime_grid_api/apps/datafile/models.py:
```diff
@@ -47,6 +48,8 @@ def get_dps(self):
 
     def generate_csv(self):
         dps = self.get_dps()
+        if len(dps) == 0:
```
For querysets, better to use `count` (https://docs.djangoproject.com/en/1.8/ref/models/querysets/#count) or `exists` (https://docs.djangoproject.com/en/1.8/ref/models/querysets/#exists) than `len(dps)`

aschn commented on watttime_grid_api/apps/datafile/tests/test_tasks.py:
```diff
@@ -80,3 +87,17 @@ def test_task_then_view(self):
+    def test_task_create_all(self):
+        csv_create_all(self.now.year, self.now.month)
+        datafiles = DataFile.objects
+        self.assertNotEqual(len(datafiles.all()), 0)
```
Could use `count` here too. Less important for the efficiency gain, but good to get in the habit :)

aschn commented:
Looks good! Ready to merge, except for the `len(queryset)` comments. You can go ahead and accept the PR after you make those little changes :)
